### ISSUE TITLE
Checksums

### ISSUE THREAD

**Issue by DemiMarie** [stratis-storage/stratisd#3031]
This will likely need to be handled by enhancements to dm-thin, or by the (still out of tree) dm-x target.  Using dm-integrity would work, but it cannot guard against replay or rollback and is slow.

**Comment by DemiMarie** [comment]
dm-integrity has two major flaws:

1. dm-integrity only provides integrity on a block-by-block basis.  This is sufficient for protecting against accidental corruption, but _not_ against malicious corruption.  An attacker can roll back the state of part of the block device and it will not be detected.
2. To avoid possible data corruption, dm-integrity requires journaling of both data and metadata, which is a significant performance penalty.

Modern storage systems such as ZFS, btrfs, and bcachefs avoid both of these problems because they are copy-on-write, with the entire filesystem a single (modulo bugs) Merkle tree rooted at the superblock.  If a collision-resistant hash function or strong MAC is used, a given superblock value uniquely identifies the entire filesystem state.  This makes partial-rollback and partial-replay attacks impossible.  Furthermore, since data is not overwritten in-place, zoned storage can be supported efficiently and journaling of data is not necessary.

There is a device-mapper target (dm-X) that provides some these features, but it is out of tree.  Furthermore, I am not sure if it can provide performance comparable to an integrated solution that uses the _other_ features CoW can provide.

**Comment by jbaublitz** [comment]
We are unlikely to use an out of tree solution, but would be open to evaluating dm-X if it ever gets merged into the mainline kernel. Our current discussions are centered around dm-integrity largely because there seems to be some integration with dm-raid which allow automatic repair of corrupted blocks in the RAID array. This would be really beneficial for us. It may be worth having a discussion with the dm-thin maintainers about whether they can support your use case specifically around the benefits of CoW combined with integrity, though! I can also bring it up the next time I talk with them just to let them know you're thinking about it if you'd like.

**Comment by DemiMarie** [comment]
It is also worth noting that cryptsetup’s AEAD options are all broken, as they use 96-bit random nonces.  Those are not long enough; you need at least 192-bit.

**Comment by DemiMarie** [comment]
> We are unlikely to use an out of tree solution, but would be open to evaluating dm-X if it ever gets merged into the mainline kernel. Our current discussions are centered around dm-integrity largely because there seems to be some integration with dm-raid which allow automatic repair of corrupted blocks in the RAID array.

Is this just that dm-integrity returns `-EIO` and dm-raid just overwrites the block to repair it?  If so, dm-X would have the same result.

> This would be really beneficial for us. It may be worth having a discussion with the dm-thin maintainers about whether they can support your use case specifically around the benefits of CoW combined with integrity, though! I can also bring it up the next time I talk with them just to let them know you're thinking about it if you'd like.

That would be awesome!  FYI the reason that CoW and integrity work so well together is that I believe temporal integrity requires CoW anyway, and multiple layers of CoW are inefficient.  This would be an always-CoW mode for dm-thin.

**Comment by jbaublitz** [comment]
> > We are unlikely to use an out of tree solution, but would be open to evaluating dm-X if it ever gets merged into the mainline kernel. Our current discussions are centered around dm-integrity largely because there seems to be some integration with dm-raid which allow automatic repair of corrupted blocks in the RAID array.
> 
> Is this just that dm-integrity returns `-EIO` and dm-raid just overwrites the block to repair it? If so, dm-X would have the same result.

Okay, great! As I understand it, that is what the interaction between dm-integrity and dm-raid is so we'll keep an eye on dm-X just in case it gets merged later.

> 
> > This would be really beneficial for us. It may be worth having a discussion with the dm-thin maintainers about whether they can support your use case specifically around the benefits of CoW combined with integrity, though! I can also bring it up the next time I talk with them just to let them know you're thinking about it if you'd like.
> 
> That would be awesome! FYI the reason that CoW and integrity work so well together is that I believe temporal integrity requires CoW anyway, and multiple layers of CoW are inefficient. This would be an always-CoW mode for dm-thin.

Okay great! I'll talk with @bmr-cymru as he may have a better idea of what the feasibility of this is.

**Comment by jbaublitz** [comment]
Would something like fs-verity support in XFS and dm-integrity together accomplish something similar to what you're looking for?

**Comment by DemiMarie** [comment]
Nope.  dm-integrity is fine at detecting accidental corruption or spacial integrity violations, but it cannot defend against replay attacks on e.g. metadata.

**Comment by jbaublitz** [comment]
More what I'm asking about is whether fs-verity gives you what you are looking for. My understanding is that btrfs (a filesystem you referenced as having the desirable Merkle tree property) utilizes fs-verity for this from what I've seen. If XFS were to implement fs-verity support as well, would that bring it in line with the aspects of btrfs that you find favorable?

**Comment by DemiMarie** [comment]
> More what I'm asking about is whether fs-verity gives you what you are looking for.

fs-verity is read-only and does not protect filesystem metadata.

> My understanding is that btrfs (a filesystem you referenced as having the desirable Merkle tree property) utilizes fs-verity for this from what I've seen.

BTRFS’s checksums are included in each data pointer and are not related to fs-verity.

> If XFS were to implement fs-verity support as well, would that bring it in line with the aspects of btrfs that you find favorable?

No, for the reasons mentioned above.  The kind of checksum support I am referring to is incompatible with in-place overwrites in the absence of data journaling.  Stratis will need a full-CoW mode anyway to support zoned storage, and checksums can be added to that mode.
